url4c27590fbe_Hoten_Java_Delaunay_tgz-pJ8-com_hoten_delaunay_examples_TestDriverJ8/src/com/hoten/delaunay/voronoi/nodename/as3delaunay/Voronoi.java

```diff
@@ -0,0 +1,484 @@
+     * index values; see PlanePointsCanvas::fillRegions()
+     * @param x
+     * @param y
+     * @return coordinates of nearest Site to (x, y)
+     *
+     */
+    /*public Point nearestSitePoint(proximityMap:BitmapData,double x, double y)
+     {
+     return _sites.nearestSitePoint(proximityMap, x, y);
+     }*/
+    public ArrayList<Point> siteCoords() {
+        return _sites.siteCoords();
+    }
+
+    private void fortunesAlgorithm() {
+        Site newSite, bottomSite, topSite, tempSite;
+        Vertex v, vertex;
+        Point newintstar = null;
+        LR leftRight;
+        Halfedge lbnd, rbnd, llbnd, rrbnd, bisector;
+        Edge edge;
+
+        Rectangle dataBounds = _sites.getSitesBounds();
+
+        int sqrt_nsites = (int) Math.sqrt(_sites.get_length() + 4);
+        HalfedgePriorityQueue heap = new HalfedgePriorityQueue(dataBounds.y, dataBounds.height, sqrt_nsites);
+        EdgeList edgeList = new EdgeList(dataBounds.x, dataBounds.width, sqrt_nsites);
+        ArrayList<Halfedge> halfEdges = new ArrayList();
+        ArrayList<Vertex> vertices = new ArrayList();
+
+        Site bottomMostSite = _sites.next();
+        newSite = _sites.next();
+
+        for (;;) {
+            if (heap.empty() == false) {
+                newintstar = heap.min();
+            }
+
+            if (newSite != null
+                    && (heap.empty() || compareByYThenX(newSite, newintstar) < 0)) {
+                /* new site is smallest */
+                //trace("smallest: new site " + newSite);
+
+                // Step 8:
+                lbnd = edgeList.edgeListLeftNeighbor(newSite.get_coord());	// the Halfedge just to the left of newSite
+                //trace("lbnd: " + lbnd);
+                rbnd = lbnd.edgeListRightNeighbor;		// the Halfedge just to the right
+                //trace("rbnd: " + rbnd);
+                bottomSite = rightRegion(lbnd, bottomMostSite);		// this is the same as leftRegion(rbnd)
+                // this Site determines the region containing the new site
+                //trace("new Site is in region of existing site: " + bottomSite);
+
+                // Step 9:
+                edge = Edge.createBisectingEdge(bottomSite, newSite);
+                //trace("new edge: " + edge);
+                _edges.add(edge);
+
+                bisector = Halfedge.create(edge, LR.LEFT);
+                halfEdges.add(bisector);
+                // inserting two Halfedges into edgeList constitutes Step 10:
+                // insert bisector to the right of lbnd:
+                edgeList.insert(lbnd, bisector);
+
+                // first half of Step 11:
+                if ((vertex = Vertex.intersect(lbnd, bisector)) != null) {
+                    vertices.add(vertex);
+                    heap.remove(lbnd);
+                    lbnd.vertex = vertex;
+                    lbnd.ystar = vertex.get_y() + newSite.dist(vertex);
+                    heap.insert(lbnd);
+                }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 341:347, 397:403.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.